index.html

```diff
@@ -174,95 +181,125 @@
-          <a>fragment</a> values.
+          The steps to <dfn data-lt="validate|valid">validate a URL-based
+          payment method identifier</dfn> are given by the following algorithm.
+          The algorithm takes a <var>potential URL</var> <a>DOMString</a> as
+          input and returns either true or an error:
         </p>
+        <ol class="algorithm">
+          <li>Let <var>url</var> be the result of trying to <a data-lt=
+          "URL Parser">parse</a> <var>potential URL</var>.
+          </li>
+          <li>If failure, <a>throw</a> a <code>TypeError</code> exception.
+          </li>
+          <li>If running <a>potentially trustworthy URL</a> with <var>url</var>
+          returns <code>"Not Trustworthy"</code>, <a>throw</a> a
+          <code>SecurityError</code> exception.
+          </li>
+          <li>If <var>url</var>'s <a>query</a> is not null or <var>url</var>'s
+          <a>fragment</a> is not null, <a>throw</a> <code>TypeError</code>
+          exception.
+          </li>
+          <li>Return true.
+          </li>
+        </ol>
       </section>
       <section>
         <h2>
           Matching
         </h2>
         <p>
-          User agents MUST use the following algorithm to determine whether two
-          payment method identifiers that are URLs match:
+          The steps to <dfn>determine whether two payment method identifiers
+          match</dfn> is given by the following algorithm. The algorithm takes
+          two string <var>A</var> and <var>B</var>.
         </p>
-        <ul>
-          <li>Let <em>A</em> be the first payment method identifier string and
-          let <em>B</em> be the second payment method identifier string.
-          </li>
-          <li>Let <em>urlA</em> be the result from the <a>URL parser</a> when
-          parsing <em>A</em> with no <a>base URL</a>.
+        <ol class="algorithm">
+          <li>Try to <a>parse</a> <var>A</var> and let <var>urlA</var> be the
+          result.
           </li>
-          <li>If either <em>urlA.query</em> or <em>urlA.fragment</em> are not
-          <em>null</em> terminate the algorithm with a result of <em>no
-          match</em> and discard <em>A</em> from further matching.
+          <li>Try to <a>parse</a> <var>B</var> and let <var>urlB</var> be the
+          result.
           </li>
-          <li>Let <em>urlB</em> be the result from the <a>URL parser</a> when
-          parsing <em>B</em> with no <a>base URL</a>.
+          <li>If either <var>urlA</var> or <var>urlB</var> is failure, return
+          false.
           </li>
-          <li>If either <em>urlB.query</em> or <em>urlB.fragment</em> are not
-          <em>null</em> terminate the algorithm with a result of <em>no
-          match</em> and discard <em>B</em> from further matching.
+          <li>
+            <a>Serialize</a> <var>urlA</var> and <var>urlB</var> and attempt to
+            <a>validate</a> each. If either throws, return false.
```

Comment:
I was hoping you'd abstract validate a bit somehow instead. Parse -> serialize -> parse seems really hacky.